app/assets/javascripts/libs/diffable_map.js

```diff
@@ -0,0 +1,294 @@
+// @flow
+// import Utils from "libs/utils";
+
+const defaultItemsPerBatch = 10000;
+let idCounter = 0;
+
+// Some properties:
+// Insertion order is not guaranteed
+
+class DiffableMap<K: number, V> {
+  id: number;
+  chunks: Array<Map<K, V>>;
+  entryCount: number;
+  existsCache: Map<K, boolean>;
+  itemsPerBatch: number;
+
+  constructor(optKeyValueArray?: ?Array<[K, V]>, itemsPerBatch: number = defaultItemsPerBatch) {
+    // Make the id property not enumerable so that it does not interfere with tests
+    Object.defineProperty(this, "id", { value: idCounter++, writable: true });
+    // this.id = idCounter++;
+    this.chunks = [];
+    this.existsCache = new Map();
+    this.entryCount = 0;
+    this.itemsPerBatch = itemsPerBatch;
+
+    if (optKeyValueArray != null) {
+      for (const [key, value] of optKeyValueArray) {
+        this.mutableSet(key, value);
+      }
+    }
+  }
+
+  get(key: K): V {
+    let idx = 0;
+    while (this.chunks[idx] != null) {
+      if (this.chunks[idx].has(key)) {
+        // $FlowFixMe
+        return this.chunks[idx].get(key);
+      }
+      idx++;
+    }
+    throw new Error("Get empty");
+  }
+
+  has(key: K): boolean {
+    try {
+      this.get(key);
+      return true;
+    } catch (exception) {
+      return false;
+    }
+  }
+
+  set(key: K, value: V): DiffableMap<K, V> {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          const newMap = shallowCopy(this);
+          newMap.chunks[idx] = new Map(this.chunks[idx]);
+          newMap.chunks[idx].set(key, value);
+          return newMap;
+        }
+        idx++;
+      }
+      throw new Error("should not happen");
+    } else {
+      const isTooFull = this.entryCount / this.chunks.length > this.itemsPerBatch;
+      const nonFullMapIdx =
+        isTooFull || this.chunks.length === 0 ? -1 : Math.floor(Math.random() * this.chunks.length);
+
+      // Key didn't exist. Add it.
+      const newDiffableMap = shallowCopy(this);
+      newDiffableMap.existsCache.set(key, true);
```

Comment:
So every key is saved twice, correct? Once in the `existsCache` and a second time in the `chunk`? Does this cost much memory? Perhaps since all values for the `existsCache` are `true` it may reference these in a saving manor.

Reply:
Yes, essentially we have two HashMaps which both map from node id's. The actual nodes are only stored in memory once. I didn't measure the memory impact, but in the end it is a memory-speed-trade-off. I also thought about using the existsCache to store the chunk index in which a node is stored to speed up the actual look up. We can tweak this in future iterations, but for now I'd like to focus on reliability.